### PR TITLE
WL-286 Add support for multi-tenancy

### PR DESCRIPTION
**Pull request by douglashall**
We need to be able to configure SOCIAL_AUTH settings per site.

Related PR:

https://github.com/edx/ecommerce/pull/543

### CODE REVIEW

**Comment by douglashall** [comment]
@clintonb @rlucioni This is ready for review.

**Comment by clintonb** [comment]
@douglashall I noticed that the tag v0.1.4 links to a commit on this branch. Tags/releases should be made after merging.

**Comment by douglashall** [comment]
@clintonb Cool. I wasn't sure of your tagging process. I can update after we merge.

File: auth_backends/backends.py [hunk 1]
```diff
@@ -2,11 +2,18 @@
 
 For more information visit https://docs.djangoproject.com/en/dev/topics/auth/customizing/.
 """
+import datetime
```
Comment (clintonb) [inline auth_backends/backends.py]: We follow the [PEP8 style for imports](https://www.python.org/dev/peps/pep-0008/#imports). Imports should be grouped as follows:

1. standard library imports
2. related third party imports
3. local application/library specific imports

Each group should be alphabetized, and you should put a blank line between each group of imports.

File: auth_backends/backends.py [hunk 2]
```diff
@@ -59,10 +86,43 @@ def auth_complete(self, *args, **kwargs):
...
+    def validate_and_return_id_token(self, id_token):
+        """
+        Validates the id_token according to the steps at
+        http://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation.
+        """
+        id_token = self._jwt_decode(id_token)
+
+        # Verify the token was issued in the last 10 minutes
+        utc_timestamp = timegm(datetime.datetime.utcnow().utctimetuple())
+        if id_token['iat'] < (utc_timestamp - 600):
```
Comment (clintonb) [inline auth_backends/backends.py]: This value should not be hard-coded. It should be configurable (via setting) for the provider.

**Comment by clintonb** [comment]
These changes should be applied to a new auth backend (e.g. `EdXOpenIdConnectMultiSite`. Overloading `EdXOpenIdConnect` leads to issues with backwards-incompatibility.

**Comment by douglashall** [comment]
Can you tell me more about the backwards-incompatibility issues you foresee?

File: auth_backends/backends.py (hunk 2 above)
Comment (douglashall) [inline auth_backends/backends.py]: This would be an enhancement that should be tackled in a separate PR.

File: auth_backends/backends.py (hunk 1 above)
Comment (douglashall) [inline auth_backends/backends.py]: Sounds good. Should be all set.

**Comment by clintonb** [comment]
If I were to drop this code into Otto today, without changing settings, would it work? If someone is sub-classing `EdXOpenIdConnect`, and decide to upgrade, will they have issues? It seems cleaner to me to make a new subclass so that it is obvious which backend is being used and which settings need to be updated. The hybrid approach will lead to confusion.

File: auth_backends/_version.py [hunk 3]
```diff
@@ -1 +1 @@
-__version__ = '0.1.3'  # pragma: no cover
+__version__ = '0.1.4'  # pragma: no cover
```
Comment (clintonb) [inline auth_backends/_version.py]: [Semantic versioning](http://semver.org/): 0.2.0

**Comment by douglashall** [comment]
I guess I'm not clear on who exactly is using this library. Seems like it would just be us and we should be in control of updating the settings of consumers. However, I will defer and create a separate backend in order to move this along.

File: auth_backends/strategies.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+from django.conf import settings
+
+from social.strategies.django_strategy import DjangoStrategy
+
+
+class CurrentSiteDjangoStrategy(DjangoStrategy):
```
Comment (clintonb) [inline auth_backends/strategies.py]: @douglashall thinking about dependencies...the `SOCIAL_AUTH_STRATEGY` can accept any Python path. That means the strategy doesn't necessarily need to live in this repo if we are concerned about dependency injection/contamination/etc. The strategy could live in edx/ecommerce, and you need only update `EdXOpenIdConnect` to use the new methods.

Comment (douglashall) [inline auth_backends/strategies.py]: Good point. Yes, I think I will move this to ecommerce. Any suggestions for where to put it? ecommerce.oauth.strategies, maybe?

File: auth_backends/backends.py [hunk 5]
```diff
@@ -99,6 +106,169 @@ def _map_user_details(self, response):
...
+        'family_name': u'last_name',
+        'locale': u'language',
+    }
+
+    auth_complete_signal = django_dispatch.Signal(providing_args=["user", "id_token"])
+
+    def id_token_issuers(self):
+        return self.setting('ISSUERS', None)
+
+    def authorization_url(self):
```
Comment (clintonb) [inline auth_backends/backends.py]: I will eventually file a PR against omab/python-social-auth to use these methods instead of the default class attributes.

File: auth_backends/strategies.py (hunk 4 above)
Comment (clintonb) [inline auth_backends/strategies.py]: s/oauth/social_auth

Comment (douglashall) [inline auth_backends/strategies.py]: Sounds good. Thanks.

Comment (clintonb) [inline auth_backends/strategies.py]: You'll want to look at python-social-auth for ideas regarding testing. The method is pretty straightforward, so no need to go crazy beyond just making sure we handle existing settings and properly fail when a setting doesn't exist.

File: auth_backends/backends.py (hunk 5 above)
Comment (douglashall) [inline auth_backends/backends.py]: That would be great. We should be able to remove this class entirely when that happens.

**Comment by douglashall** [comment]
@clintonb This is ready for a final review. We will need to get this merged and get the 0.2.0 version pushed to pypi before the tests will run for https://github.com/edx/ecommerce/pull/543.

File: auth_backends/backends.py [hunk 6]
```diff
@@ -99,6 +103,169 @@ def _map_user_details(self, response):
         return dest
 
 
+class EdXSettingsAwareOpenIdConnect(OpenIdConnectAuth):
```
Comment (clintonb) [inline auth_backends/backends.py]: Why not inherit from `EdXOpenIdConnect`? Doing so would get rid of a lot of duplicated code.

File: auth_backends/backends.py [hunk 7]
```diff
@@ -99,6 +103,169 @@ def _map_user_details(self, response):
...
+        'email': u'email',
+        'name': u'full_name',
+        'given_name': u'first_name',
+        'family_name': u'last_name',
+        'locale': u'language',
+    }
+
+    auth_complete_signal = django_dispatch.Signal(providing_args=["user", "id_token"])
+
+    def id_token_issuers(self):
```
Comment (clintonb) [inline auth_backends/backends.py]: Why is this pluralized? There should only be one issuer.

Comment (douglashall) [inline auth_backends/backends.py]: I believe the spec allows for multiple issuers. http://openid.net/specs/openid-connect-core-1_0.html#IssuerIdentifier

File: auth_backends/backends.py (hunk 6 above)
Comment (douglashall) [inline auth_backends/backends.py]: All set.

**Comment by clintonb** [comment]
The current implementation is fine aside from the unnecessary duplication. Thinking back to another option we discussed, I think this could be simplified even further:

```python
class EdXOpenIdConnect(OpenIdConnectAuth):
  @property
  def ID_TOKEN_ISSUER(self)
    return self.setting('URL_ROOT')

  @property
  def AUTHORIZATION_URL(self)
    return '{0}/authorize/'.format(self.ID_TOKEN_ISSUER)
```
This approach would maintain backwards-compatibility while enabling a new strategy to support site-specific settings.

**Comment by douglashall** [comment]
@clintonb Alright, I think we have made this as concise as possible. Thanks for your help on this.

File: auth_backends/backends.py [hunk 8]
```diff
@@ -99,6 +99,24 @@ def _map_user_details(self, response):
         return dest
 
 
+class EdXSettingsAwareOpenIdConnect(EdXOpenIdConnect):
```
Comment (clintonb) [inline auth_backends/backends.py]: I believe these changes can be applied directly to `EdXOpenIdConnect`. There is no longer a need for the child class since we have backwards-compatibility.

Comment (clintonb) [inline auth_backends/backends.py]: Also, tests are needed for these new properties.

**Comment by clintonb** [comment]
:+1:, pending final comments addressed.

**Comment by douglashall** [comment]
@mattdrayer Could you give this a second review?

File: auth_backends/backends.py [hunk 9]
```diff
@@ -34,6 +30,22 @@ class EdXOpenIdConnect(OpenIdConnectAuth):
 
     auth_complete_signal = django.dispatch.Signal(providing_args=["user", "id_token"])
 
+    @property
+    def ID_TOKEN_ISSUER(self):
+        return self.setting('ISSUER')
```
Comment (clintonb) [inline auth_backends/backends.py]: Note that the consuming service will need to have this value set (in addition to URL_ROOT). This value should be set in configuration.

Comment (clintonb) [inline auth_backends/backends.py]: It seems this advice was never heeded, and now I have one more thing to fix for LEARNER-693. Come on!

**Comment by douglashall** [comment]
Hi Clinton,

Sincere apologies if this is causing you pain. I really appreciate all of the great work that you do and do not want to make it difficult for you to continue that work. My hope is to learn all I can from you.

It’s not clear to me from this comment what the problem is or what needs to be done to solve it. Would you mind describing the problem you are running into in more detail and the solution which you are implementing so that we can avoid pain for everyone in the future?

Sorry again and thanks,

- Doug

**Comment by clintonb** [comment]
Sorry for lashing out.

There is no "ISSUER" setting as far as I can tell. I'm not sure how/why
logins were working in the past. Now login is broken on
credentials.stage.edx.org because the setting cannot be found. Was the
issuer ever configured?

Clinton Blackburn

**Comment by douglashall** [comment]
No worries. Is this “ISSUER” setting in edxapp or credentials? Sorry, trying to refresh my memory about how this all works.

**Comment by clintonb** [comment]
Both. edxapp, as the issuer, needs to know what value to set. All of the IDAs need to know what value to expect.
